### PR TITLE
Cascaded shadow mapping

### PR DESCRIPTION
**Pull request by mvaligursky**
Implementation of stable cascaded shadow mapping for directional lights:
- 1, 2, 3 or 4 cascades (1 is not cascaded, it's a reimplementation of existing shadow mapping using unified code).
- all shadow map filtering modes work (PCF3, PCF5 and VSM)

**New API:**
- **`LightComponent#numCascades`** - Number of cascades
- **`LightComponent#cascadeDistribution`** - The distribution of subdivision of the camera frustum for individual shadow cascades.

Fixes: https://github.com/playcanvas/engine/issues/2937
Possible enhancements: https://github.com/playcanvas/engine/issues/3193

Example: http://playcanvas.github.io/#graphics/shadow-cascades.html

Screenshot 2021-05-25 at 11 01 22

### CODE REVIEW

File: src/graphics/program-lib/chunks/shadowCascades.frag [hunk 1]
```diff
@@ -0,0 +1,54 @@
...
+void FadeShadow(float shadowCascadeDistances[4]) {                  
+
+    // if the pixel is past the shadow distance, remove shadow
+    // this enforces straight line instead of corner of shadow which moves when camera rotates  
+    float depth = 1.0 / gl_FragCoord.w;
+    if (depth > shadowCascadeDistances[int(maxCascades - 1.0)]) {
+        dShadowCoord.z = -9999999.0;
+    }
+}
+
```
Comment (willeastcott) [inline src/graphics/program-lib/chunks/shadowCascades.frag:54]:
```suggestion

```

File: src/graphics/program-lib/chunks/shadowCascades.frag [hunk 2]
```diff
@@ -0,0 +1,53 @@
+const float maxCascades = 4.0;
+
+// shadow matrix for selected cascade
+mat4 cascadeShadowMat;
+
+// funtion which based on cascade distances select a shadow projection matrix to use
```
Comment (willeastcott) [inline src/graphics/program-lib/chunks/shadowCascades.frag:6]:
```suggestion
// function which selects a shadow projection matrix based on cascade distances 
```

File: src/framework/components/light/component.js [hunk 3]
```diff
@@ -334,6 +339,12 @@ function _defineProps() {
     _defineProperty("shadowBias", 0.05, function (newValue, oldValue) {
         this.light.shadowBias = -0.01 * newValue;
     });
+    _defineProperty("shadowCascades", 1, function (newValue, oldValue) {
+        this.light.numCascades = newValue;
```
Comment (willeastcott) [inline src/framework/components/light/component.js:343]: It is worth some sanitizing of the parameter here? Or a DEBUG message when it's not 1, 2, 3 or 4?

File: src/graphics/program-lib/chunks/shadowCascades.frag [hunk 4]
```diff
@@ -0,0 +1,53 @@
...
+void getShadowCascadeMatrix(mat4 shadowMatrixPalette[4], float shadowCascadeDistances[4], float shadowCascadeCount) {
+
+    // depth in 0 .. far plane range
+    float depth = 1.0 / gl_FragCoord.w;
+
+    // find cascade index based on the depth (loop as there is no per component vec compare operator in webgl)
+    float cascadeIndex = 0.0;
+    for (float i = 0.0; i < maxCascades; i++) {
+        if (depth < shadowCascadeDistances[int(i)])
+        {
```
Comment (willeastcott) [inline src/graphics/program-lib/chunks/shadowCascades.frag:16]:
```suggestion
        if (depth < shadowCascadeDistances[int(i)]) {
```

File: src/graphics/program-lib/chunks/shadowCascades.frag [hunk 5]
```diff
@@ -0,0 +1,53 @@
...
+        else if (cascadeIndex == 2.0) {
+            cascadeShadowMat = shadowMatrixPalette[2];
+        }
+        else {
+            cascadeShadowMat = shadowMatrixPalette[3];
+        }
+    #endif
+}
+
+void FadeShadow(float shadowCascadeDistances[4]) {
```
Comment (willeastcott) [inline src/graphics/program-lib/chunks/shadowCascades.frag:45]: Is the convention to use camel case for GLSL function naming in our codebase?

File: src/math/mat4.js [hunk 6]
```diff
@@ -738,6 +738,45 @@ class Mat4 {
...
+     * to their position inside a viewport (in range of 0 .. 1). This encapsulates a scaling to the size of the viewport
+     * and a translation to the position of the viewport.
+     * @param {number} x - The x-component of the position of the viewport (in 0..1 range).
+     * @param {number} y - The y-component of the position of the viewport (in 0..1 range).
+     * @param {number} width - The width of the viewport (in 0..1 range).
+     * @param {number} height - The height of the viewport (in 0..1 range).
+     * @returns {Mat4} Self for chaining.
+     * @example
+     * // Create a 4x4 scale matrix
+     * var sm = new pc.Mat4().setScale(10, 10, 10);
```
Comment (willeastcott) [inline src/math/mat4.js:755]: Example code needs updating.

File: src/scene/renderer/shadow-renderer.js [hunk 7]
```diff
@@ -321,118 +299,218 @@ class ShadowRenderer {
...
+            const x = Math.ceil(center.dot(up) * sizeRatio) / sizeRatio;
+            const y = Math.ceil(center.dot(right) * sizeRatio) / sizeRatio;
+
+            const scaledUp = up.mulScalar(x);
+            const scaledRight = right.mulScalar(y);
+            const dot = center.dot(lightDir);
+            const scaledDir = lightDir.mulScalar(dot);
+            center.add2(scaledUp, scaledRight).add(scaledDir);
+
+            // look at the center from war away to include all casters during culling
```
Comment (willeastcott) [inline src/scene/renderer/shadow-renderer.js:339]:
```suggestion
            // look at the center from far away to include all casters during culling
```

File: src/scene/renderer/shadow-renderer.js [hunk 8]
```diff
@@ -321,118 +299,218 @@ class ShadowRenderer {
...
                 }
             }
 
-            visibleCasters.length = count;
+            // calculate depth range of the caster's AABB from the point of view of the shadow camera
+            shadowCamView.copy(shadowCamNode.getWorldTransform()).invert();
+            const depthRange = getDepthRange(shadowCamView, visibleSceneAabb.getMin(), visibleSceneAabb.getMax());
+
+            // adjust shadow camera's near and far plane to the depth range of casters to maximize precision
+            // of values stores in the shadow map
```
Comment (willeastcott) [inline src/scene/renderer/shadow-renderer.js:369]:
```suggestion
            // of values stored in the shadow map
```

File: src/math/mat4.js (hunk 6 above)
Comment (mvaligursky) [reply src/math/mat4.js]: nice one, done.

File: src/graphics/program-lib/chunks/shadowCascades.frag (hunk 5 above)
Comment (mvaligursky) [reply src/graphics/program-lib/chunks/shadowCascades.frag]: done, thanks

File: src/framework/components/light/component.js (hunk 3 above)
Comment (mvaligursky) [reply src/framework/components/light/component.js]: good idea and done, thanks

File: examples/graphics/lights.html [hunk 9]
```diff
@@ -146,10 +147,11 @@
                 castShadows: true,
                 intensity: 0.6
             });
-            omnilight.addComponent("model", {
-                type: "sphere"
+            omnilight.addComponent("render", {
+                type: "sphere",
+                castShadows: false,
+                material: createMaterial({ diffuse: new pc.Color(0, 0, 0), emissive: pc.Color.YELLOW })
```
Comment (willeastcott) [inline examples/graphics/lights.html:153]:
```suggestion
                material: createMaterial({ diffuse: new pc.Color.BLACK, emissive: pc.Color.YELLOW })
```

File: src/framework/components/light/component.js [hunk 10]
```diff
@@ -334,6 +339,12 @@ function _defineProps() {
     _defineProperty("shadowBias", 0.05, function (newValue, oldValue) {
         this.light.shadowBias = -0.01 * newValue;
     });
+    _defineProperty("numCascades", 1, function (newValue, oldValue) {
+        this.light.numCascades = math.clamp(Math.floor(newValue), 1, 4);
+    });
+    _defineProperty("cascadeDistribution", 0.5, function (newValue, oldValue) {
+        this.light.cascadeDistribution = newValue;
```
Comment (willeastcott) [inline src/framework/components/light/component.js:346]:
```suggestion
        this.light.cascadeDistribution = math.clamp(newValue, 0, 1);
````

**Review by willeastcott — APPROVED** [review]

File: src/scene/light.js [hunk 11]
```diff
@@ -393,7 +405,8 @@ class Light {
         // 16 - 17 : cookie channel G
         // 14 - 15 : cookie channel B
         // 12      : cookie transform
-        //  9 - 11 : light source shape
+        // 10 - 11 : light source shape
+        //  7 -  9 : light num cascades
```
Comment (willeastcott) [inline src/scene/light.js:409]: Couldn't this fit in 2 bits if you minus 1 from `numCascades`? Or is that not worth saving?

Comment (mvaligursky) [reply src/scene/light.js]: it can never hurt to save a bit. done, thanks!

**Pull request merged**